### PR TITLE
Optimizer: enable the DeinitDevirtualizer pass

### PR DESCRIPTION
**Pull request by eeckstein**
It was disabled so far to not hide bugs in the deinit code generation. Now hopefully deinit code generation is stable enough to enable the pass.

Also fix a bug some bugs which showed up after enabling the pass.
E.g. don't convert non-copyable owned -> guaranteed arguments in FunctionSignatureOpts: It's not safe to insert a compensating release_value at the call site. This release_value calls the deinit which might have been already de-virtualized in the callee.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

File: test/IRGen/raw_layout_multifile.swift [hunk 1]
```diff
@@ -1,6 +1,6 @@
 // RUN: %empty-directory(%t)
 // RUN: %{python} %utils/chex.py < %s > %t/raw_layout_multifile.swift
-// RUN: %target-swift-frontend -enable-experimental-feature BuiltinModule -enable-experimental-feature RawLayout -emit-ir -O -disable-availability-checking %s %S/Inputs/raw_layout_multifile_b.swift | %FileCheck %t/raw_layout_multifile.swift --check-prefix=CHECK --check-prefix=CHECK-%target-ptrsize
+// RUN: %target-swift-frontend -enable-experimental-feature BuiltinModule -enable-experimental-feature RawLayout -emit-ir -O -Xllvm -sil-disable-pass=deinit-devirtualizer -disable-availability-checking %s %S/Inputs/raw_layout_multifile_b.swift | %FileCheck %t/raw_layout_multifile.swift --check-prefix=CHECK --check-prefix=CHECK-%target-ptrsize
```
Comment (Azoy) [inline test/IRGen/raw_layout_multifile.swift:3]: I'm curious why you had to disable this pass here 🤔 Does raw layout not work nicely with deinit devirt?

Comment (eeckstein) [reply test/IRGen/raw_layout_multifile.swift:3]: No. If deinits are devirtualized, it looks like that no value witness table is generated at all. And this is checked in the test

**Review by atrick — APPROVED** [review]

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Pull request merged**